pkg/events/transport/kafka.go

```diff
@@ -5,4 +5,198 @@
+	}
+}
+
+// NewKafka creates an instance of transport based on Kafka broker.
+func NewKafka(brokerAddrs []string, options ...func(k *Kafka) error) (*Kafka, error) {
+	syncProducer, err := sarama.NewSyncProducer(brokerAddrs, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	k := Kafka{
+		producer: syncProducer,
+	}
+	for _, option := range options {
+		// TODO: handle errors from options
+		option(&k)
+	}
+	if k.producerOnly {
+		return &k, nil
+	}
+
+	k.consumer, err = sarama.NewConsumer(brokerAddrs, nil)
+	return &k, err
+}
+
+// Publish publishes an event
+func (k *Kafka) Publish(ctx context.Context, event *events.CloudEvent, topic string, tenant string) error {
+	defer trace.Tracef("topic: %s", event.EventType)()
+	sp, _ := opentracing.StartSpanFromContext(
+		ctx,
+		"Kafka.Publish",
+	)
+	defer sp.Finish()
+
+	msg, err := fromEvent(event)
+	if err != nil {
+		return errors.Wrapf(err, "error when creating Kafka message from CloudEvent for topic %s", topic)
+	}
+	msg.Topic = topic
+
+	err = injectSpan(sp, msg)
+	if err != nil {
+		return errors.Wrap(err, "error injecting opentracing span to Kafka message")
+	}
+
+	if _, _, err = k.producer.SendMessage(msg); err != nil {
+		return errors.Wrap(err, "error sending Kafka message")
+	}
+
+	return nil
+}
+
+// Subscribe subscribes to an event
+func (k *Kafka) Subscribe(ctx context.Context, topic string, handler events.Handler) (events.Subscription, error) {
+	defer trace.Tracef("topic: %s", topic)()
+	sp, _ := opentracing.StartSpanFromContext(
+		ctx,
+		"Kafka.Subscribe",
+	)
+	defer sp.Finish()
+
+	doneChan := make(chan struct{})
+
+	// create partition consumer. Since we are creating only one consumer, we should messages from all partitions
+	// regardless of the partition number we select
+	partitionConsumer, err := k.consumer.ConsumePartition(topic, 0, sarama.OffsetNewest)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating a partition consumer for topic %s", topic)
+	}
+
+	go func() {
+		defer trace.Tracef("listening for messages on topic: %s", topic)()
+		for {
+			select {
+			case msg, open := <-partitionConsumer.Messages():
+				if !open {
+					partitionConsumer.Close()
+					return
+				}
+				spCtx, _ := extractSpan(msg)
+				spSub := opentracing.StartSpan(
+					"Kafka.SubscriptionHandler",
+					opentracing.FollowsFrom(spCtx),
+				)
+				func() {
+					defer spSub.Finish()
+					// Update the context with the span for the subsequent reference.
+					ctx = opentracing.ContextWithSpan(context.Background(), spSub)
+					event, err := toEvent(msg)
+					if err != nil {
+						log.Errorf("Error when converting Kafka message to event: %+v", err)
+						return
+					}
+					log.Debugf("Got an event %+v", event)
+					handler(ctx, event)
+				}()
+
+			case <-doneChan:
+				partitionConsumer.Close()
+				return
+			}
+		}
+	}()
+
+	return &subscription{done: doneChan}, nil
+}
+
+// Close closes the transport
+func (k *Kafka) Close() {
+	err := k.producer.Close()
```

Comment:
Do you need to close the consumer as well?

Reply:
good catch, will fix that too